Implement 'extract' command

Xinayder opened the pull request:
The only thing missing is to actually extract the files, but it requires a bit more of work.

Xinayder committed:
Implement 'extract' command

phrohdoh commented on cli-tools/sagebig/src/main.rs:
```diff
@@ -97,3 +113,32 @@ fn cmd_contains(args: &clap::ArgMatches) -> Result<(), ReadError> {
+            .expect(&format!("Failed to read known entry {} from {}", name, big_path));
+
+        let file_path = output_dir.join(Path::new(&entry.name));
+        let file_dirpath = &file_path.parent().unwrap();
```
Please don't unwrap.

This could be carried out in the root of a fs and this would panic.

phrohdoh commented on cli-tools/sagebig/src/main.rs:
```diff
@@ -97,3 +113,32 @@ fn cmd_contains(args: &clap::ArgMatches) -> Result<(), ReadError> {
+        let file_dirpath = &file_path.parent().unwrap();
+
+        if !file_dirpath.exists() {
+            fs::DirBuilder::new().recursive(true).create(&file_dirpath).unwrap();
```
Same thing here regarding unwrap.

This should be `try!`-ed with an `impl From<SomeErrorType>` as necessary.

phrohdoh replied:
Thinking about this a bit more we don't want this to return a `ReadError` because this isn't only reading.

We'll need some sort of overarching, geneirc `libbig::Error` that contains Read/Write/etc

phrohdoh commented on cli-tools/sagebig/src/main.rs:
```diff
@@ -97,3 +113,32 @@ fn cmd_contains(args: &clap::ArgMatches) -> Result<(), ReadError> {
+
+        if !file_dirpath.exists() {
+            fs::DirBuilder::new().recursive(true).create(&file_dirpath).unwrap();
+            println!("Created path `{}`", &file_dirpath.display());
```
Would you kindly put these `println!`s behind a `-v|--verbose` flag?

hint: https://docs.rs/clap/2.20.0/clap/struct.ArgMatches.html#method.is_present

phrohdoh reviewed (changes requested):
I've left some inline comments requesting changes, please let me know when these have been addresses so this can be merged!

RockyTV commented:
```rust
From 92b24ac780672d908c45c867121720f4b169db41 Mon Sep 17 00:00:00 2001
From: Alexandre Oliveira <rockytvbr@gmail.com>
Date: Wed, 18 Jan 2017 15:15:01 -0200
Subject: [PATCH] Stop unwrapping, add verbose arg

---
 cli-tools/sagebig/src/main.rs | 21 ++++++++++++++++-----
 1 file changed, 16 insertions(+), 5 deletions(-)

diff --git a/cli-tools/sagebig/src/main.rs b/cli-tools/sagebig/src/main.rs
index d27f5e9..bed4da2 100644
--- a/cli-tools/sagebig/src/main.rs
+++ b/cli-tools/sagebig/src/main.rs
@@ -56,7 +56,9 @@ fn main() {
             .arg(Arg::with_name("output_dir")
                 .value_name("output_dir")
                 .required(true)
-                .index(2)))
+                .index(2))
+            .arg(Arg::with_name("verbose")
+                .short("v")))
         .get_matches();
 
     let res = match matches.subcommand() {
@@ -118,11 +120,15 @@ fn cmd_extract(args: &clap::ArgMatches) -> Result<(), ReadError> {
     let big_path = args.value_of("archive_path").unwrap();
     let archive = try!(BigArchive::new_from_path(&big_path));
     let output_dir = Path::new(args.value_of("output_dir").unwrap());
+    let verbose = args.is_present("verbose");
 
     // Create output dir if it doesn't exist
     if !output_dir.exists() {
         try!(fs::create_dir(&output_dir));
-        println!("Created output_dir `{:?}`", &output_dir.display());
+
+        if verbose {
+            println!("Created output_dir `{:?}`", &output_dir.display());
+        }
     }
 
     let output_dir = output_dir.canonicalize().unwrap();
@@ -132,11 +138,16 @@ fn cmd_extract(args: &clap::ArgMatches) -> Result<(), ReadError> {
             .expect(&format!("Failed to read known entry {} from {}", name, big_path));
 
         let file_path = output_dir.join(Path::new(&entry.name));
-        let file_dirpath = &file_path.parent().unwrap();
+        let file_dirpath = match file_path.parent() {
+            Some(x) => x,
+            None => panic!(format!("Invalid file path detected: {}", &file_path.display()))
+        };
 
         if !file_dirpath.exists() {
-            fs::DirBuilder::new().recursive(true).create(&file_dirpath).unwrap();
-            println!("Created path `{}`", &file_dirpath.display());
+            try!(fs::DirBuilder::new().recursive(true).create(&file_dirpath));
+            if verbose {
+                println!("Created path `{}`", &file_dirpath.display());
+            }
         }
     }
 
-- 
2.10.2


```

Xinayder committed:
Add new Error enum

Errors were moved to libbig::errors module. There's also a new error
type: ExtractError.

phrohdoh commented on cli-tools/sagebig/src/main.rs:
```diff
@@ -74,7 +93,7 @@ fn cmd_list(args: &clap::ArgMatches) -> Result<(), ReadError> {
 }
 
-fn cmd_search(args: &clap::ArgMatches) -> Result<(), ReadError> {
+fn cmd_search(args: &clap::ArgMatches) -> Result<(), Error> {
```
This will never encounter a write error so should remain `ReadError`.

phrohdoh commented on cli-tools/sagebig/src/main.rs:
```diff
@@ -61,7 +80,7 @@ fn main() {
 }
 
-fn cmd_list(args: &clap::ArgMatches) -> Result<(), ReadError> {
+fn cmd_list(args: &clap::ArgMatches) -> Result<(), Error> {
```
Same here.

phrohdoh commented on cli-tools/sagebig/src/main.rs:
```diff
@@ -88,7 +107,7 @@ fn cmd_search(args: &clap::ArgMatches) -> Result<(), ReadError> {
 }
 
-fn cmd_contains(args: &clap::ArgMatches) -> Result<(), ReadError> {
+fn cmd_contains(args: &clap::ArgMatches) -> Result<(), Error> {
```
Same here.

phrohdoh commented on cli-tools/sagebig/src/main.rs:
```diff
@@ -97,3 +116,41 @@ fn cmd_contains(args: &clap::ArgMatches) -> Result<(), ReadError> {
+        }
+    }
+
+    let output_dir = output_dir.canonicalize().unwrap();
```
No unwrap!

phrohdoh commented on src/errors/mod.rs:
```diff
@@ -0,0 +1,45 @@
+
+#[derive(Debug)]
+pub enum ExtractError {
+    IoError(io::Error),
```
Please name this `StdIoError` for consistency with `ReadError`.

phrohdoh commented on src/errors/mod.rs:
```diff
@@ -0,0 +1,45 @@
+pub enum Error {
+    ReadError(ReadError),
+    ExtractError(ExtractError),
+    IoError(io::Error)
```
Is this necessary?
Read and Extract both contain an `std::io::Error` variant.

Phrohdoh commented:
With ae46a7207e4f76536d7bbc4f7e6896cd5ec15060 you should be able to implement file writing too, but please don't take too much advantage of getting a `Vec<u8>` (as opposed to a `&[u8]`) I hope to change that as soon as I can.

phrohdoh commented on cli-tools/sagebig/src/main.rs:
```diff
@@ -97,3 +115,44 @@ fn cmd_contains(args: &clap::ArgMatches) -> Result<(), ReadError> {
+        let file_path = output_dir.join(Path::new(&entry.name));
+        let file_dirpath = match file_path.parent() {
+            Some(x) => x,
+            None => return Err(Error::ExtractError(ExtractError::InvalidFilePath(file_path .to_string_lossy().into_owned()))),
```
I think what you want is [`Path::is_file`](https://doc.rust-lang.org/nightly/std/path/struct.Path.html#method.is_file) paired with a possible `Path::parent`.

I say possible because `foo.txt` in the root is completely valid.

Phrohdoh commented:
Closing this along with #2, thank you for your work!
